src/main/scala/org/broadinstitute/hail/stats/ComputeRRM.scala

```diff
@@ -0,0 +1,82 @@
+package org.broadinstitute.hail.stats
+
+import breeze.linalg._
+import org.apache.spark.mllib.linalg.distributed.{IndexedRow, IndexedRowMatrix, RowMatrix}
+import org.apache.spark.mllib.linalg.{Vectors, DenseMatrix => SDenseMatrix}
+import org.broadinstitute.hail.utils._
+import org.broadinstitute.hail.variant.{Genotype, VariantDataset}
+
+// mean 0, norm sqrt(n), variance 1 (constant variants are None)
+object ToNormalizedGtArray {
+  def apply(gs: Iterable[Genotype], nSamples: Int): Option[Array[Double]] = {
+    val (nPresent, gtSum, gtSumSq) = gs.flatMap(_.gt).foldLeft((0, 0, 0))((acc, gt) => (acc._1 + 1, acc._2 + gt, acc._3 + gt * gt))
+    val nMissing = nSamples - nPresent
+    val allHomRef = gtSum == 0
+    val allHet = gtSum == nPresent && gtSumSq == nPresent
+    val allHomVar = gtSum == 2 * nPresent
+
+    if (allHomRef || allHomVar || allHet)
+      None
+    else {
+      val gtMean = gtSum.toDouble / nPresent
+      val gtMeanAll = (gtSum + nMissing * gtMean) / nSamples
+      val gtMeanSqAll = (gtSumSq + nMissing * gtMean * gtMean) / nSamples
+      val gtStdDevRec = 1d / math.sqrt(gtMeanSqAll - gtMeanAll * gtMeanAll)
+
+      Some(gs.map(_.gt.map(g => (g - gtMean) * gtStdDevRec).getOrElse(0d)).toArray)
+    }
+  }
+}
+
+// each row has mean 0, norm sqrt(n), variance 1 (constant variants are dropped)
+object ToNormalizedRowMatrix {
+  def apply(vds: VariantDataset): RowMatrix = {
+    val n = vds.nSamples
+    val rows = vds.rdd.flatMap { case (v, (va, gs)) => ToNormalizedGtArray(gs, n) }.map(Vectors.dense)
+    val m = rows.count()
+    new RowMatrix(rows, m, n)
+  }
+}
+
+// each row has mean 0, norm sqrt(n), variance 1, constant variants are dropped
+object ToNormalizedIndexedRowMatrix {
+  def apply(vds: VariantDataset): IndexedRowMatrix = {
+    val n = vds.nSamples
+    val variants = vds.variants.collect()
+    val variantIdxBc = vds.sparkContext.broadcast(variants.index)
+    val indexedRows = vds.rdd.flatMap { case (v, (va, gs)) => ToNormalizedGtArray(gs, n).map(a => IndexedRow(variantIdxBc.value(v), Vectors.dense(a))) }
+    new IndexedRowMatrix(indexedRows, variants.length, n)
+  }
+}
+
+// diagonal values are approximately m assuming independent variants by Central Limit Theorem
+object ComputeLocalGrammian {
+  def withoutBlock(A: RowMatrix): DenseMatrix[Double] = {
+    val n = A.numCols().toInt
+    val G = A.computeGramianMatrix().asInstanceOf[SDenseMatrix].values
```

Comment:
I changed them to toArray. These are currently only called once in context of far bigger computation, so avoiding a copy on a local array is not worth violating API safety.